db/compaction/compaction_job.cc

```diff
@@ -1655,15 +2168,30 @@ void CompactionJob::ProcessKeyValueCompaction(SubcompactionState* sub_compact) {
     status = c_iter->status();
   }
 
-  if (status.ok() && sub_compact->builder == nullptr &&
-      sub_compact->outputs.size() == 0 && !range_del_agg.IsEmpty()) {
+  if (status.ok() && !sub_compact->HasBuilder() && !sub_compact->HasOutput() &&
+      !range_del_agg.IsEmpty()) {
     // handle subcompaction containing only range deletions
     status = OpenCompactionOutputFile(sub_compact);
   }
 
   // Call FinishCompactionOutputFile() even if status is not ok: it needs to
   // close the output file.
-  if (sub_compact->builder != nullptr) {
+  // If it has precluded bottommost outputs, close them first.
+  if (sub_compact->HasPrecludedBottommostOutputs()) {
+    sub_compact->OverrideCurrentOutputs(true);
+    if (sub_compact->HasBuilder()) {
+      CompactionIterationStats range_del_out_stats;
+      Status s = FinishCompactionOutputFile(status, sub_compact, &range_del_agg,
+                                            &range_del_out_stats);
+      if (!s.ok() && status.ok()) {
+        status = s;
+      }
+      RecordDroppedKeys(range_del_out_stats,
+                        &sub_compact->compaction_job_stats);
+    }
+    sub_compact->OverrideCurrentOutputs(false);
+  }
```

Comment:
Did we consider to just make output files an array of 1 or 2, based on whether a preclude level is needed? In this way, code like this can be a for-loop, and it's harder to forget some places we treat output files.

Reply:
I tried that at the first, but all caller needs to handle that (with either 2 for loops or nested for loops). I would prefer hiding that information within `SubcompactionState`.